File: go/cmd/tiny-go/telemetry.go
package main

import (
	"context"
	"strings"
	"sync"
	"time"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
)

type openTelemetryMonitor struct {
	mu         sync.Mutex
	provider   *sdktrace.TracerProvider
	tracer     trace.Tracer
	startup    trace.Span
	operations map[string]trace.Span
	models     map[string]trace.Span
	tools      map[string]trace.Span
	mcp        map[string]trace.Span
	model      string
}

func newOpenTelemetryMonitor(env map[string]string) lifecycleSink {
	if strings.EqualFold(env["OTEL_SDK_DISABLED"], "true") || (env["OTEL_EXPORTER_OTLP_ENDPOINT"] == "" && env["OTEL_EXPORTER_OTLP_TRACES_ENDPOINT"] == "") {
		return callbackLifecycleSink(func(map[string]any) {})
	}
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	exporter, err := otlptracehttp.New(ctx)
	if err != nil {
		return callbackLifecycleSink(func(map[string]any) {})
	}
	serviceName := env["OTEL_SERVICE_NAME"]
	if serviceName == "" {
		serviceName = "tiny-go"
	}
	base, err := resource.Merge(resource.Default(), resource.NewWithAttributes("", attribute.String("service.name", serviceName), attribute.String("openinference.project.name", serviceName)))
	if err != nil {
		return callbackLifecycleSink(func(map[string]any) {})
	}
	provider := sdktrace.NewTracerProvider(sdktrace.WithResource(base), sdktrace.WithBatcher(exporter))
	return &openTelemetryMonitor{
		provider: provider, tracer: provider.Tracer("tiny-agent"),
		operations: map[string]trace.Span{}, models: map[string]trace.Span{}, tools: map[string]trace.Span{}, mcp: map[string]trace.Span{},
	}
}

func (m *openTelemetryMonitor) Emit(event map[string]any) {
	m.mu.Lock()
	defer m.mu.Unlock()
	timestamp := lifecycleTime(event["timestamp"])
	switch stringValue(event["type"]) {
	case "startup.started":
		if m.startup != nil {
			endTelemetrySpan(m.startup, "effect_unknown", timestamp)
		}
		m.model = stringValue(event["model"])
		_, m.startup = m.tracer.Start(context.Background(), "tiny.startup", trace.WithTimestamp(timestamp), trace.WithAttributes(
			attribute.String("tiny.runtime.language", stringValue(event["runtime"])),
			attribute.Int("tiny.plugin.count", sliceLength(event["plugins"])),
			attribute.Int("tiny.mcp.server_count", sliceLength(event["mcp"])),
		))
	case "mcp.started":
		if m.startup == nil {
			return
		}
		server := stringValue(event["server"])
		_, span := m.tracer.Start(trace.ContextWithSpan(context.Background(), m.startup), "tiny.mcp.connect", trace.WithTimestamp(timestamp), trace.WithAttributes(attribute.String("tiny.mcp.server", server)))
		m.mcp[server] = span
	case "mcp.completed":
		server := stringValue(event["server"])
		span := m.mcp[server]
		if span == nil {
			return
		}
		setTelemetryAttributes(span, map[string]any{"tiny.mcp.protocol_version": event["protocolVersion"], "tiny.mcp.tool_count": event["toolCount"], "error.type": event["errorType"]})
		endTelemetrySpan(span, stringValue(event["outcome"]), timestamp)
		delete(m.mcp, server)
	case "startup.completed":
		if m.startup == nil {
			return
		}
		setTelemetryAttributes(m.startup, map[string]any{"error.type": event["errorType"]})
		endTelemetrySpan(m.startup, stringValue(event["outcome"]), timestamp)
		m.startup = nil
	case "operation.started", "operation.recovered":
		id := stringValue(event["operationId"])
		if old := m.operations[id]; old != nil {
			endTelemetrySpan(old, "effect_unknown", timestamp)
		}
		_, span := m.tracer.Start(context.Background(), "tiny.agent.operation", trace.WithTimestamp(timestamp), trace.WithAttributes(telemetryAttributes(map[string]any{
			"openinference.span.kind": "AGENT", "gen_ai.operation.name": "invoke_agent", "gen_ai.conversation.id": event["sessionId"], "session.id": event["sessionId"],
			"tiny.session.id": event["sessionId"], "tiny.operation.id": id, "tiny.operation.kind": event["operationKind"], "tiny.recovery": event["recovery"], "tiny.runtime.language": "go",
		})...))
		m.operations[id] = span
	case "model.started":
		parent := m.operations[stringValue(event["operationId"])]
		if parent == nil {
			return
		}
		id := stringValue(event["attemptId"])
		_, span := m.tracer.Start(trace.ContextWithSpan(context.Background(), parent), "tiny.model.request", trace.WithTimestamp(timestamp), trace.WithAttributes(telemetryAttributes(map[string]any{
			"openinference.span.kind": "LLM", "gen_ai.operation.name": "chat", "gen_ai.conversation.id": event["sessionId"], "gen_ai.request.model": m.model, "llm.model_name": m.model, "session.id": event["sessionId"],
			"tiny.session.id": event["sessionId"], "tiny.operation.id": event["operationId"], "tiny.operation.kind": event["operationKind"], "tiny.step.id": event["stepId"], "tiny.attempt.id": id, "tiny.attempt.number": event["attempt"], "tiny.recovery": event["recovery"],
		})...))
		m.models[id] = span
	case "model.completed":
		id := stringValue(event["attemptId"])
		span := m.models[id]
		if span == nil {
			return
		}
		setTelemetryAttributes(span, usageTelemetryAttributes(event["usage"]))
		setTelemetryAttributes(span, map[string]any{"error.type": event["errorType"]})
		endTelemetrySpan(span, stringValue(event["outcome"]), timestamp)
		delete(m.models, id)
	case "model.reconciled", "tool.admitted", "cancel.requested":
		span := m.operations[stringValue(event["operationId"])]
		if span != nil {
			span.AddEvent(stringValue(event["type"]), trace.WithTimestamp(timestamp), trace.WithAttributes(telemetryAttributes(event)...))
		}
	case "tool.started":
		parent := m.operations[stringValue(event["operationId"])]
		if parent == nil {
			return
		}
		id := stringValue(event["attemptId"])
		_, span := m.tracer.Start(trace.ContextWithSpan(context.Background(), parent), "tiny.tool.execute", trace.WithTimestamp(timestamp), trace.WithAttributes(telemetryAttributes(map[string]any{
			"openinference.span.kind": "TOOL", "gen_ai.operation.name": "execute_tool", "gen_ai.conversation.id": event["sessionId"], "gen_ai.tool.call.id": event["toolCallId"], "gen_ai.tool.name": event["tool"], "session.id": event["sessionId"], "tool.name": event["tool"],
			"tiny.session.id": event["sessionId"], "tiny.operation.id": event["operationId"], "tiny.step.id": event["stepId"], "tiny.attempt.id": id, "tiny.parent_attempt.id": event["parentAttemptId"], "tiny.tool.started.id": event["toolStartedId"], "tiny.recovery": event["recovery"],
		})...))
		m.tools[id] = span
	case "tool.completed":
		id := stringValue(event["attemptId"])
		span := m.tools[id]
		if span == nil {
			return
		}
		endTelemetrySpan(span, stringValue(event["outcome"]), timestamp)
		delete(m.tools, id)
	case "operation.completed":
		id := stringValue(event["operationId"])
		span := m.operations[id]
		if span == nil {
			return
		}
		setTelemetryAttributes(span, map[string]any{"tiny.operation.completion": event["completion"], "error.type": event["errorType"]})
		endTelemetrySpan(span, stringValue(event["outcome"]), timestamp)
		delete(m.operations, id)
	}
}

func (m *openTelemetryMonitor) Close() error {
	m.mu.Lock()
	timestamp := time.Now()
	if m.startup != nil {
		endTelemetrySpan(m.startup, "effect_unknown", timestamp)
	}
	for _, spans := range []map[string]trace.Span{m.mcp, m.models, m.tools, m.operations} {
		for _, span := range spans {
			endTelemetrySpan(span, "effect_unknown", timestamp)
		}
	}
	m.mu.Unlock()
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	return m.provider.Shutdown(ctx)
}

func usageTelemetryAttributes(value any) map[string]any {
	usage, ok := lifecycleUsage(value)
	if !ok {
		return nil
	}
	input := usage.Input + usage.CacheRead + usage.CacheWrite
	return map[string]any{
		"gen_ai.usage.input_tokens": input, "gen_ai.usage.output_tokens": usage.Output,
		"gen_ai.usage.cache_read.input_tokens": usage.CacheRead, "gen_ai.usage.cache_creation.input_tokens": usage.CacheWrite,
		"llm.token_count.prompt": input, "llm.token_count.completion": usage.Output, "llm.token_count.total": input + usage.Output,
		"tiny.usage.input_tokens": usage.Input, "tiny.usage.cache_read_tokens": usage.CacheRead, "tiny.usage.cache_write_tokens": usage.CacheWrite,
	}
}

func setTelemetryAttributes(span trace.Span, values map[string]any) {
	span.SetAttributes(telemetryAttributes(values)...)
}
func telemetryAttributes(values map[string]any) []attribute.KeyValue {
	result := make([]attribute.KeyValue, 0, len(values))
	for key, value := range values {
		switch value := value.(type) {
		case string:
			if value != "" {
				result = append(result, attribute.String(key, value))
			}
		case bool:
			result = append(result, attribute.Bool(key, value))
		case int:
			result = append(result, attribute.Int(key, value))
		case int64:
			result = append(result, attribute.Int64(key, value))
		case float64:
			result = append(result, attribute.Float64(key, value))
		}
	}
	return result
}
func endTelemetrySpan(span trace.Span, outcome string, timestamp time.Time) {
	span.SetAttributes(attribute.String("tiny.outcome", outcome))
	if outcome == "succeeded" {
		span.SetStatus(codes.Ok, "")
	} else if outcome != "cancelled" {
		span.SetStatus(codes.Error, outcome)
	}
	span.End(trace.WithTimestamp(timestamp))
}
func sliceLength(value any) int {
	if values, ok := value.([]string); ok {
		return len(values)
	}
	if values, ok := value.([]any); ok {
		return len(values)
	}
	return 0
}

File: python/tiny_agent/telemetry.py
import os
from datetime import datetime, timezone

from opentelemetry import trace
from opentelemetry.context import Context
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.trace import Span, Status, StatusCode


class NoTelemetry:
    def emit(self, event: dict) -> None:
        pass

    def close(self) -> None:
        pass


class OpenTelemetryMonitor:
    def __init__(self, tracer, shutdown=lambda: None, runtime: str = "python"):
        self.tracer, self.shutdown, self.runtime = tracer, shutdown, runtime
        self.startup: tuple[Span, Context] | None = None
        self.operations: dict[str, tuple[Span, Context]] = {}
        self.models: dict[str, Span] = {}
        self.tools: dict[str, Span] = {}
        self.mcp: dict[str, Span] = {}
        self.model: str | None = None

    def emit(self, event: dict) -> None:
        try:
            self._record(event)
        except Exception:
            pass

    def close(self) -> None:
        for span in [*self.mcp.values(), *self.models.values(), *self.tools.values(), *(item[0] for item in self.operations.values())]:
            _end(span, "effect_unknown")
        if self.startup:
            _end(self.startup[0], "effect_unknown")
        self.mcp.clear()
        self.models.clear()
        self.tools.clear()
        self.operations.clear()
        self.startup = None
        try:
            self.shutdown()
        except Exception:
            pass

    def _record(self, event: dict) -> None:
        kind = event.get("type")
        if kind == "startup.started":
            if self.startup:
                _end(self.startup[0], "effect_unknown")
            self.model = event.get("model")
            span = self.tracer.start_span(
                "tiny.startup",
                start_time=_event_time(event),
                attributes={
                    "tiny.runtime.language": self.runtime,
                    "tiny.plugin.count": len(event.get("plugins", [])),
                    "tiny.mcp.server_count": len(event.get("mcp", [])),
                },
            )
            self.startup = span, trace.set_span_in_context(span)
            return
        if kind == "mcp.started":
            context = self.startup[1] if self.startup else None
            self.mcp[event["server"]] = self.tracer.start_span("tiny.mcp.connect", context=context, start_time=_event_time(event), attributes={"tiny.mcp.server": event["server"]})
            return
        if kind == "mcp.completed":
            span = self.mcp.pop(event.get("server"), None)
            if not span:
                return
            span.set_attributes(_attributes({"tiny.mcp.protocol_version": event.get("protocolVersion"), "tiny.mcp.tool_count": event.get("toolCount"), "error.type": event.get("errorType")}))
            _end(span, event.get("outcome", "failed"), _event_time(event))
            return
        if kind == "startup.completed":
            if not self.startup:
                return
            span, _ = self.startup
            self.startup = None
            span.set_attributes(_attributes({"error.type": event.get("errorType")}))
            _end(span, event.get("outcome", "failed"), _event_time(event))
            return
        if kind in ("operation.started", "operation.recovered"):
            operation_id = event["operationId"]
            existing = self.operations.pop(operation_id, None)
            if existing:
                _end(existing[0], "effect_unknown")
            span = self.tracer.start_span(
                "tiny.agent.operation",
                start_time=_event_time(event),
                attributes={
                    "openinference.span.kind": "AGENT",
                    "gen_ai.operation.name": "invoke_agent",
                    "gen_ai.conversation.id": event["sessionId"],
                    "session.id": event["sessionId"],
                    "tiny.session.id": event["sessionId"],
                    "tiny.operation.id": operation_id,
                    "tiny.operation.kind": event["operationKind"],
                    "tiny.recovery": event["recovery"],
                    "tiny.runtime.language": self.runtime,
                },
            )
            self.operations[operation_id] = span, trace.set_span_in_context(span)
            return
        if kind == "model.started":
            parent = self.operations.get(event.get("operationId"))
            if not parent:
                return
            self.models[event["attemptId"]] = self.tracer.start_span(
                "tiny.model.request",
                context=parent[1],
                start_time=_event_time(event),
                attributes=_attributes(
                    {
                        "openinference.span.kind": "LLM",
                        "gen_ai.operation.name": "chat",
                        "gen_ai.conversation.id": event["sessionId"],
                        "gen_ai.request.model": self.model,
                        "llm.model_name": self.model,
                        "session.id": event["sessionId"],
                        "tiny.session.id": event["sessionId"],
                        "tiny.operation.id": event["operationId"],
                        "tiny.operation.kind": event["operationKind"],
                        "tiny.step.id": event["stepId"],
                        "tiny.attempt.id": event["attemptId"],
                        "tiny.attempt.number": event["attempt"],
                        "tiny.recovery": event["recovery"],
                    }
                ),
            )
            return
        if kind == "model.completed":
            span = self.models.pop(event.get("attemptId"), None)
            if not span:
                return
            span.set_attributes(_attributes({**_usage_attributes(event.get("usage")), "error.type": event.get("errorType")}))
            _end(span, event.get("outcome", "failed"), _event_time(event))
            return
        if kind in ("model.reconciled", "tool.admitted", "cancel.requested"):
            operation = self.operations.get(event.get("operationId"))
            if not operation:
                return
            values = (
                {"tiny.step.id": event.get("stepId"), "tiny.attempt.id": event.get("attemptId"), "tiny.outcome": event.get("outcome"), "error.type": event.get("errorType")}
                if kind == "model.reconciled"
                else {
                    "tiny.step.id": event.get("stepId"),
                    "tiny.tool.started.id": event.get("toolStartedId"),
                    "gen_ai.tool.call.id": event.get("toolCallId"),
                    "gen_ai.tool.name": event.get("tool"),
                    "tiny.tool.replay": event.get("replay"),
                }
                if kind == "tool.admitted"
                else {"tiny.cancel.phase": event.get("phase"), "gen_ai.tool.call.id": event.get("toolCallId")}
            )
            operation[0].add_event(kind, _attributes(values), _event_time(event))
            return
        if kind == "tool.started":
            parent = self.operations.get(event.get("operationId"))
            if not parent:
                return
            attempt_id = event["attemptId"]
            existing = self.tools.pop(attempt_id, None)
            if existing:
                _end(existing, "effect_unknown")
            self.tools[attempt_id] = self.tracer.start_span(
                "tiny.tool.execute",
                context=parent[1],
                start_time=_event_time(event),
                attributes={
                    "openinference.span.kind": "TOOL",
                    "gen_ai.operation.name": "execute_tool",
                    "gen_ai.conversation.id": event["sessionId"],
                    "gen_ai.tool.call.id": event["toolCallId"],
                    "gen_ai.tool.name": event["tool"],
                    "session.id": event["sessionId"],
                    "tool.name": event["tool"],
                    "tiny.session.id": event["sessionId"],
                    "tiny.operation.id": event["operationId"],
                    "tiny.step.id": event["stepId"],
                    "tiny.attempt.id": attempt_id,
                    "tiny.parent_attempt.id": event["parentAttemptId"],
                    "tiny.tool.started.id": event["toolStartedId"],
                    "tiny.recovery": event["recovery"],
                },
            )
            return
        if kind == "tool.completed":
            span = self.tools.pop(event.get("attemptId"), None)
            if span:
                _end(span, event.get("outcome", "failed"), _event_time(event))
            return
        if kind == "operation.completed":
            operation = self.operations.pop(event.get("operationId"), None)
            if not operation:
                return
            operation[0].set_attributes(_attributes({"tiny.operation.completion": event.get("completion"), "error.type": event.get("errorType")}))
            _end(operation[0], event.get("outcome", "failed"), _event_time(event))


def create_telemetry(env: dict[str, str] | None = None):
    env = os.environ if env is None else env
    if env.get("OTEL_SDK_DISABLED", "").lower() == "true":
        return NoTelemetry()
    if not (env.get("OTEL_EXPORTER_OTLP_ENDPOINT") or env.get("OTEL_EXPORTER_OTLP_TRACES_ENDPOINT")):
        return NoTelemetry()
    try:
        service_name = env.get("OTEL_SERVICE_NAME") or "tiny-py"
        provider = TracerProvider(resource=Resource.create({"service.name": service_name, "openinference.project.name": service_name}))
        provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter()))
        return OpenTelemetryMonitor(provider.get_tracer("tiny-agent", "1"), provider.shutdown)
    except Exception:
        return NoTelemetry()


def _event_time(event: dict) -> int:
    value = event.get("timestamp")
    try:
        return int(datetime.fromisoformat(value.replace("Z", "+00:00")).timestamp() * 1_000_000_000)
    except AttributeError, ValueError:
        return int(datetime.now(timezone.utc).timestamp() * 1_000_000_000)


def _attributes(values: dict) -> dict:
    return {key: value for key, value in values.items() if isinstance(value, (str, int, float, bool))}


def _usage_attributes(usage: dict | None) -> dict:
    if not usage:
        return {}
    prompt = usage["input"] + usage["cacheRead"] + usage["cacheWrite"]
    return {
        "gen_ai.usage.input_tokens": prompt,
        "gen_ai.usage.output_tokens": usage["output"],
        "gen_ai.usage.cache_read.input_tokens": usage["cacheRead"],
        "gen_ai.usage.cache_creation.input_tokens": usage["cacheWrite"],
        "llm.token_count.prompt": prompt,
        "llm.token_count.completion": usage["output"],
        "llm.token_count.total": prompt + usage["output"],
        "tiny.usage.input_tokens": usage["input"],
        "tiny.usage.cache_read_tokens": usage["cacheRead"],
        "tiny.usage.cache_write_tokens": usage["cacheWrite"],
    }


def _end(span: Span, outcome: str, end_time: int | None = None) -> None:
    span.set_attribute("tiny.outcome", outcome)
    span.set_status(Status(StatusCode.OK if outcome == "succeeded" else StatusCode.UNSET if outcome == "cancelled" else StatusCode.ERROR))
    span.end(end_time=end_time)
